src/GraphQL/Value.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+-- | Literal GraphQL values.
+module GraphQL.Value
+  (
+    -- | GraphQL values
+    GraphQL.Value.Value(..)
+  , ToValue(..)
+  , Name
+  , List
+  , Map
+  , String
+    -- | Fields
+  , Field(Field)
+  , makeField
+  , FieldSet
+  , GraphQL.Value.empty
+  , singleton
+  , fromList
+  , fieldSetToMap
+  , union
+  , unions
+  ) where
+
+import Protolude hiding (Map)
+
+import Data.Foldable (foldrM)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import Data.GraphQL.AST (Name)
+import Data.Aeson (ToJSON(..), Value(Null))
+
+-- | Concrete GraphQL value. Essentially Data.GraphQL.AST.Value, but without
+-- the "variable" field.
+data Value
+  = ValueInt Int32
+  | ValueFloat Double
+  | ValueBoolean Bool
+  | ValueString String
+  | ValueEnum Name
+  | ValueList List
+  | ValueMap Map
+  | ValueNull
+  deriving (Eq, Ord, Show)
+
+instance ToJSON GraphQL.Value.Value where
+
+  toJSON (ValueInt x) = toJSON x
+  toJSON (ValueFloat x) = toJSON x
+  toJSON (ValueBoolean x) = toJSON x
+  toJSON (ValueString x) = toJSON x
+  toJSON (ValueEnum x) = toJSON x
+  toJSON (ValueList x) = toJSON x
+  toJSON (ValueMap x) = toJSON x
+  toJSON ValueNull = Null
+
+newtype String = String Text deriving (Eq, Ord, Show)
+
+instance ToJSON String where
+  toJSON (String x) = toJSON x
+
+newtype List = List [GraphQL.Value.Value] deriving (Eq, Ord, Show)
+
+makeList :: (Functor f, Foldable f, ToValue a) => f a -> List
+makeList = List . toList . map toValue
+
+instance ToJSON List where
+  toJSON (List x) = toJSON x
+
+-- XXX: This is ObjectValue [ObjectField]; ObjectField Name Value upstream.
+-- XXX: GraphQL spec itself sometimes says 'map' and other times 'object', but
+-- jml hasn't read 100% clearly. Let's find something and stick to it, and
+-- make sure that there isn't a real distinction between to the two.
+newtype Map = Map (Map.Map Name GraphQL.Value.Value) deriving (Eq, Ord, Show)
+
+instance ToJSON Map where
+  toJSON (Map x) = toJSON x
+
+
+data Field = Field Name GraphQL.Value.Value deriving (Eq, Show, Ord)
+
+makeField :: (StringConv name Name, ToValue value) => name -> value -> Field
+makeField name value = Field (toS name) (toValue value)
+
+data FieldSet = FieldSet (Set Field) deriving (Eq, Show)
+
+instance ToValue FieldSet where
+  toValue = toValue . fieldSetToMap
+
+fieldSetToMap :: FieldSet -> Map
+fieldSetToMap (FieldSet fields) = Map (Map.fromList [ (name, value) | Field name value <- toList fields ])
+
+empty :: FieldSet
+empty = FieldSet Set.empty
+
+singleton :: Field -> FieldSet
+singleton = FieldSet . Set.singleton
+
+-- TODO: Fail on duplicate keys.
+union :: Alternative m => FieldSet -> FieldSet -> m FieldSet
```

Comment:
is `union` related to graphql union types? If so I'm not sure why it should fail on duplicates